File: 02.01.移除重复节点/02.01.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeDuplicateNodes(ListNode* head) {
        if (head == nullptr) {
            return nullptr;
        }

        std::unordered_set<int> set;
        set.insert(head->val);

        auto pos = head;
        while (pos->next != nullptr) {
            auto cur = pos->next;
            if (set.count(cur->val)) {
                pos->next = pos->next->next;
            } else {
                set.insert(cur->val);
                pos = pos->next;
            }
        }

        return head;
    }
};

File: 02.01.移除重复节点/02.01.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func removeDuplicateNodes(head *ListNode) *ListNode {
    if head == nil {
        return nil
    }

    m := make(map[int]bool)
    m[head.Val] = true

    pos := head
    for pos.Next != nil {
        cur := pos.Next
        if m[cur.Val] {
            pos.Next = pos.Next.Next
        } else {
            m[cur.Val] = true
            pos = pos.Next
        }
    }

    return head
}